File: backend/app/schemas/blog.py
from pydantic import BaseModel
from datetime import date
from typing import List
from app.graph.blog_graph import (
    Plan,
    Task,
    EvidenceItem
)

class GenerateBlogRequest(BaseModel):
    topic: str
    as_of: date

class ImageResponse(BaseModel):
    placeholder: str
    filename: str

    alt: str
    caption: str

    prompt: str
    size: str
    quality: str

    image_url: str | None = None
    blob_path: str | None = None

class BlogResponse(BaseModel):
    topic: str
    as_of: date
    plan: Plan | None
    tasks: List[Task]
    evidence: List[EvidenceItem]
    markdown: str
    images: List[ImageResponse]
    logs: List[str]

class SaveBlogRequest(BaseModel):
    topic: str

    as_of: str

    plan: Plan | None

    tasks: list[Task]

    evidence: list[EvidenceItem] | None

    images: List[ImageResponse] | None

    markdown: str

    logs: list[str]

File: frontend/src/types/blog.ts
export interface GenerateBlogRequest {
  topic: string;
  as_of: string;
}

export interface Plan {
  blog_title: string;
  audience: string;
  tone: string;
  blog_kind: string;
  constraints: string[];
}

export interface Task {
  id: number;
  title: string;
  target_words: number;
  tags: string[];

  requires_research: boolean;
  requires_citations: boolean;
  requires_code: boolean;
}

export interface Evidence {
  title: string;
  url: string;
  published_at: string;
  snippet: string;
  source: string;
}

export interface BlogImage {
  placeholder: string;
  filename: string;
  alt: string;
  caption: string;
  prompt: string;
  image_url: string;
}

export interface BlogResponse {
  topic: string;
  as_of: string;
  
  plan: Plan | null;
  tasks: Task[];

  evidence: Evidence[];

  markdown: string;

  images: BlogImage[] | null;

  logs: string[];
}

export interface SaveResponse {
  message: string;
  id: number;
}

export interface BlogListItem {
  id: number;
  topic: string;
  blog_title: string;
  created_at: string;
}

export interface StreamStep {
  label: string;
  status:
    | "pending"
    | "running"
    | "completed"
    | "error";
}